### PR TITLE
chore(sentryapp): add regional RPC service to get request buffers by region

### PR DESCRIPTION
**Pull request by ameliahsu**
this is the first step in fixing our `SentryAppRequestsEndpoint` to return requests from all regions

added
- `find_regions_for_sentry_app` in `region.py` to get all regions with an installation for the specified SentryApp
- `regions_with_installations` wrapper method for `find_regions_for_sentry_app` in `SentryApp` model
- new `RegionApp` RPC service with `get_buffer_requests_for_region` method to fetch buffer requests for the specified SentryApp in the specified region

### CODE REVIEW

File: src/sentry/sentry_apps/services/region_app/__init__.py [hunk 1]
```diff
@@ -0,0 +1,2 @@
+from .model import *  # noqa
```
Comment (iamrajjoshi) [inline src/sentry/sentry_apps/services/region_app/__init__.py:1]: style nit: is adding these import statements here required for hybrid cloud purposes? 
if not, i'd recommend we remove them so we enforce folks to have exact imports. this helps us prevent circular dependency issues in the future.

**Review by iamrajjoshi — COMMENTED** [review]
going to let others more familiar with hybrid cloud stuff review the service, just had a nit about imports

File: src/sentry/types/region.py [hunk 2]
```diff
@@ -343,6 +346,21 @@ def find_regions_for_user(user_id: int) -> set[str]:
...
+def find_regions_for_sentry_app(sentry_app: SentryApp) -> set[str]:
+    from sentry.sentry_apps.models.sentry_app_installation import SentryAppInstallation
+
+    if SiloMode.get_current_mode() == SiloMode.MONOLITH:
+        return {settings.SENTRY_MONOLITH_REGION}
+
+    org_ids = set(
+        SentryAppInstallation.objects.filter(sentry_app=sentry_app).values_list(
+            "organization_id", flat=True
+        )
```
Comment (GabeVillalobos) [inline src/sentry/types/region.py:359]: This can potentially return 1000s of organization IDs. We may want to consider doing the aggregation and org region lookups in a single query if possible, but that means we won't be able to use the `find_regions_for_orgs` helper :/

Comment (ameliahsu) [reply src/sentry/types/region.py:359]: I tried to do this in a single query but failed 🫠  I don't think the Organization model has a foreign key linking to the OrganizationMapping model where the region info is, so I wasn't able to do one long query

File: src/sentry/types/region.py [hunk 3]
```diff
@@ -16,6 +16,9 @@
 from sentry.utils import json
 from sentry.utils.env import in_test_environment
 
+if TYPE_CHECKING:
+    from sentry.sentry_apps.models.sentry_app import SentryApp
```
[the remark this answers is not in the document]
Comment (cathteng) [reply src/sentry/types/region.py:20]: you can also import this directly before it's used

File: src/sentry/types/region.py [hunk 4]
```diff
@@ -343,6 +346,21 @@ def find_regions_for_user(user_id: int) -> set[str]:
...
 
+@control_silo_function
+def find_regions_for_sentry_app(sentry_app: SentryApp) -> set[str]:
+    from sentry.sentry_apps.models.sentry_app_installation import SentryAppInstallation
+
+    if SiloMode.get_current_mode() == SiloMode.MONOLITH:
+        return {settings.SENTRY_MONOLITH_REGION}
+
+    org_ids = set(
+        SentryAppInstallation.objects.filter(sentry_app=sentry_app).values_list(
```
Comment (cathteng) [inline src/sentry/types/region.py:357]: if this is a `set` and we're worried about pulling all of the org ids into Python, you can do this in one go with something like
`SentryAppInstallation.objects.filter(sentry_app=sentry_app).distinct('organization_id')` which will already be a set

File: src/sentry/types/region.py (hunk 2 above)
Comment (markstory) [reply src/sentry/types/region.py:359]: Organization and OrganizationMapping can't have relations because they are located in different silos. While you can't join sentry_app and organization mapping tables together, you could use a subquery

```python
installation_query = SentryAppInstallation.objects.filter(sentry_app=sentry_app).values_list("organization_id")
regions = OrganizationMapping.objects.filter(organization_id__in=installation_query).distinct("region")
```

This avoids loading the large list of organization_ids into memory and only fetches the region list which will always be  0-2 rows.

File: src/sentry/sentry_apps/services/region_app_request_buffer/service.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+from sentry.sentry_apps.services.app import RpcSentryApp
+from sentry.sentry_apps.services.region_app_request_buffer.model import (
+    RpcSentryAppRequest,
+    SentryAppRequestFilterArgs,
+)
+from sentry.silo.base import SiloMode
+
+
+class RegionAppService(RpcService):
+    key = "region_app"
```
Comment (markstory) [inline src/sentry/sentry_apps/services/region_app_request_buffer/service.py:14]: SentryAppRequestService 🤷 I think we might be better off with names that help clarify what how the data is used/collected over where it lives.

File: src/sentry/sentry_apps/services/region_app_request_buffer/service.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        )
+
+        return DatabaseBackedRegionAppService()
+
+    @regional_rpc_method(resolve=ByRegionName())
+    @abc.abstractmethod
+    def get_buffer_requests_for_region(
+        self,
+        *,
+        sentry_app: RpcSentryApp,
```
[the remark this answers is not in the document]
Comment (ameliahsu) [reply src/sentry/sentry_apps/services/region_app_request_buffer/service.py:30]: We need to pass the entire RPC object to fetch the `SentryAppRequestBuffer`

https://github.com/getsentry/sentry/blob/bcef5b4b9ce59259c1a827d8e20a2b3eb81ebbc9/src/sentry/sentry_apps/services/region_app_request_buffer/impl.py#L21

[the remark this answers is not in the document]
Comment (GabeVillalobos) [reply src/sentry/sentry_apps/services/region_app_request_buffer/service.py:30]: An alternative would be to decouple the `SentryAppWebhookRequestsBuffer` class from `sentry_app`, but it looks like it requires slightly modifying the `add_request` call which only relies on the `is_internal` property: https://github.com/getsentry/sentry/blob/c5b12fd51019e8bb483e10410bbaae34dc03b30c/src/sentry/utils/sentry_apps/request_buffer.py#L180

Simplifying this now to always just set the organization ID in `add_request`, regardless of whether this is internal or no  seems like a good way to sidestep having to move around an RPC sentry app.

File: src/sentry/sentry_apps/services/region_app/__init__.py (hunk 1 above)
Comment (markstory) [reply src/sentry/sentry_apps/services/region_app/__init__.py]: There isn't any functionality hanging off barrel files for services. However, you'll want to ensure that new services are imported during application startup. If a service isn't imported during startup RPC requests to it will fail as the service won't be part of the global registry.

File: src/sentry/sentry_apps/services/region_app_request_buffer/service.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (markstory) [reply src/sentry/sentry_apps/services/region_app_request_buffer/service.py:30]: You could have the `id` be part of the request, and in the rpc method, use that id to fetch a fresh copy of the app.

[the remark this answers is not in the document]
Comment (GabeVillalobos) [reply src/sentry/sentry_apps/services/region_app_request_buffer/service.py:30]: I'm not even sure we need a full copy of the sentry app to begin with though. The usage of it within the `SentryAppWebhookRequestsBuffer` class is already super minimal, which is why I think we could remove it entirely. Is there some other benefit to refetching the sentry app?

File: src/sentry/types/region.py (hunk 4 above)
Comment (GabeVillalobos) [reply src/sentry/types/region.py:357]: Dropping this here because I found this behavior really confusing and had to look this up, but `distinct(<fields>)` doesn't return a set, as it doesn't run a `SELECT DISTINCT(<columns>)` query. Instead, it runs a `SELECT DISTINCT ON(<fields>)` query specifically when using postgres: https://docs.djangoproject.com/en/5.1/ref/models/querysets/#django.db.models.query.QuerySet.distinct

So you still need a `values_list("region_name")` here 😅

Comment (cathteng) [reply src/sentry/types/region.py:357]: oh. i forgot this was about region name, i wrote this about the org_id part 😓

File: src/sentry/sentry_apps/services/region_app_request_buffer/service.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (markstory) [reply src/sentry/sentry_apps/services/region_app_request_buffer/service.py:30]: >  Is there some other benefit to refetching the sentry app?
 
No, it was a way to not have to change `SentryAppRequestBuffer` and avoid serializing the object into RPC parameters.

File: src/sentry/sentry_apps/services/app_request/serial.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        date=request.get("date"),
+        response_code=request.get("response_code"),
+        webhook_url=request.get("webhook_url"),
+        organization_id=request.get("organization_id"),
+        event_type=request.get("event_type"),
+        error_id=request.get("error_id"),
+        project_id=request.get("project_id"),
+        request_body=request.get("request_body"),
+        request_headers=rpc_request_headers,
+        response_body=request.get("response_body"),
```
Comment (GabeVillalobos) [inline src/sentry/sentry_apps/services/app_request/serial.py:30]: Should these have `None` as the defaults if they're optional in the `RpcSentryAppRequest`?

File: src/sentry/sentry_apps/services/app_request/model.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    date: str
+    response_code: int
+    webhook_url: str
+    organization_id: int
+    event_type: str
+    error_id: str | None = None
+    project_id: int | None = None
+    request_body: str | None = None
+    request_headers: RpcSentryAppRequestHeaders | None = None
+    response_body: str | None = None
```
Comment (GabeVillalobos) [inline src/sentry/sentry_apps/services/app_request/model.py:29]: Are we displaying this at all in the updated design? If not, does it make sense to exclude this from the model?

File: tests/sentry/sentry_apps/services/test_app_request.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+                        "settings": {
+                            "type": "alert-rule-settings",
+                            "uri": "https://example.com/search/",
+                        },
+                    },
+                ]
+            },
+        )
+
+    def test_get_buffer_requests_for_region(self):
```
Comment (GabeVillalobos) [inline tests/sentry/sentry_apps/services/test_app_request.py:34]: Nice job with testing! It may be good to have a test that also validates an empty buffer requests list.

File: tests/sentry/types/test_region.py [hunk 10]
```diff
@@ -189,6 +190,32 @@ def test_find_regions_for_user(self) -> None:
...
+                scopes=["project:write"],
+            )
+            self.create_sentry_app_installation(slug=sentry_app.slug, organization=organization)
+            actual_regions = find_regions_for_sentry_app(sentry_app=sentry_app)
+            assert actual_regions == {"us"}
+
+            eu_organization = self.create_organization(name="eu test name", region="eu")
+            self.create_sentry_app_installation(slug=sentry_app.slug, organization=eu_organization)
+            actual_regions = find_regions_for_sentry_app(sentry_app=sentry_app)
+            assert actual_regions == {"us", "eu"}
```
Comment (GabeVillalobos) [inline tests/sentry/types/test_region.py:211]: It may be good to add multiple orgs per region to ensure these dedupe correctly

File: tests/sentry/types/test_region.py [hunk 11]
```diff
@@ -189,6 +190,32 @@ def test_find_regions_for_user(self) -> None:
...
+
+        with (
+            override_settings(SILO_MODE=SiloMode.REGION),
+            pytest.raises(SiloLimit.AvailabilityError),
+        ):
+            find_regions_for_sentry_app(sentry_app=sentry_app)
+
```
Comment (GabeVillalobos) [inline tests/sentry/types/test_region.py:218]: Also a test for empty regions list would be good

File: src/sentry/sentry_apps/services/app_request/serial.py (hunk 7 above)
Comment (GabeVillalobos) [reply src/sentry/sentry_apps/services/app_request/serial.py:30]: Nvm, this is covered by get already.

**Review by GabeVillalobos — APPROVED** [review]

File: src/sentry/sentry_apps/services/app_request/service.py [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+from sentry.hybridcloud.rpc.service import RpcService, regional_rpc_method
+from sentry.sentry_apps.services.app_request.model import (
+    RpcSentryAppRequest,
+    SentryAppRequestFilterArgs,
+)
+from sentry.silo.base import SiloMode
+
+
+class SentryAppRequestService(RpcService):
+    key = "region_app"
```
Comment (markstory) [inline src/sentry/sentry_apps/services/app_request/service.py:13]:
```suggestion
    key = "sentry_app_request"
```

This name is used in metrics and URLs. Having it line up with the classname is an informal convention.

**Review by markstory — APPROVED** [review]
Looks good other than a small naming update.
